motion_computation/src/bsm_to_external_object_convertor.cpp

```diff
@@ -0,0 +1,348 @@
+  } 
+  else if (in_msg.core_data.accuracy.presence_vector |
+          carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_ORIENTATION_AVAILABLE) 
+  {
+    // Orientation accuracy available
+
+    lat_variance = 1.0;
+    lon_variance = 1.0;
+  }
+  // Else: No accuracies available
+
+  /////// Pose and Covariance /////////
+  // Compute the pose
+
+  if ((in_msg.core_data.presence_vector & carma_v2x_msgs::msg::BSMCoreData::LATITUDE_AVAILABLE) && 
+      (in_msg.core_data.presence_vector & carma_v2x_msgs::msg::BSMCoreData::LONGITUDE_AVAILABLE) && 
+      (in_msg.core_data.presence_vector & carma_v2x_msgs::msg::BSMCoreData::ELEVATION_AVAILABLE) && 
+      (in_msg.core_data.presence_vector & carma_v2x_msgs::msg::BSMCoreData::HEADING_AVAILABLE))
+  {
+    double longitude = std::min(in_msg.core_data.longitude * 0.0000001, in_msg.core_data.LONGITUDE_MAX);
+    longitude = std::max(longitude, in_msg.core_data.LONGITUDE_MIN);
+    
+    double latitude =  std::min(in_msg.core_data.latitude * 0.0000001, in_msg.core_data.LATITUDE_MAX);
+    latitude =  std::max(latitude, in_msg.core_data.LATITUDE_MIN);
+    
+    double elev = std::min(in_msg.core_data.elev * 0.1, (double)in_msg.core_data.ELEVATION_MAX);
+    elev = std::max(elev, (double)in_msg.core_data.ELEVATION_MIN);
+
+    double heading = std::min(in_msg.core_data.heading * 0.0125, (double)in_msg.core_data.HEADING_MAX);
+    heading = std::max(heading, (double)in_msg.core_data.HEADING_MIN);
+
+    out_msg.pose = impl::pose_from_gnss(map_projector, ned_in_map_rotation,
+                                      {latitude, longitude, elev},
+                                      heading, lat_variance, lon_variance, heading_variance);
+    out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::POSE_PRESENCE_VECTOR;
+  }
+  // Else: No pose available
+
+  
+  out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::POSE_PRESENCE_VECTOR;
+
+  /////// Predictions /////////
+  // Compute predictions
+  // For prediction, if the prediction is available we will sample it
+  // If not then assume linear motion
+
+  std::vector<geometry_msgs::msg::Pose> predicted_poses;
+
+  predicted_poses = impl::sample_2d_linear_motion(out_msg.pose.pose, out_msg.velocity.twist.linear.x, pred_period, pred_step_size);
+
+  out_msg.predictions = impl::predicted_poses_to_predicted_state(predicted_poses, out_msg.velocity.twist.linear.x, rclcpp::Time(out_msg.header.stamp),
+                                                                 rclcpp::Duration(pred_step_size * 1e9), map_frame_id, out_msg.confidence,
+                                                                 out_msg.confidence);
+  out_msg.presence_vector |= carma_perception_msgs::msg::ExternalObject::PREDICTION_PRESENCE_VECTOR;
+}
+
+namespace impl {
+
+std::vector<geometry_msgs::msg::Pose> sample_2d_linear_motion(const geometry_msgs::msg::Pose& pose, double velocity,
```

Comment:
Rather than duplicating this function, can you just include the psm helper and call it directly